### PR TITLE
Introducing Visualizer (SandDance) to the SQL Query Editor.

### PR DESCRIPTION
**Pull request by rchlkm**
This is the first pull request by Garage interns working on the data visualization experience in Azure Data Studio using SandDance.

The garageMVP branch is up to date with the master branch. garageMVP also has changes from [alanren/ExtRecommendationAppLaunch](https://github.com/microsoft/azuredatastudio/pull/6105). Alan Ren's code structure was used to create the Visualizer (SandDance) extension recommendation popup. 

--
We introduced a Visualizer icon in the SQL Query Editor Results Grid. This icon will be reviewed soon to ensure it meets the icon standards. Currently, we have both the normal and inverse icons.

visualizer_gridactions

viewVisualizer

When a user clicks on this icon, if they do not have the SandDance extension, they will be prompted to install the extension. Since the extension is required to utilize the feature, they will not have the option to "Don't show again".
visualizerIcon_popup

"Install Extension" will download the SandDance extension. When more visualizer extensions are added, the same button can be used to install all the visualizer extensions at once as long as product.json is updated.

"More info" will take the user to the extension marketplace that shows all the visualizer extensions. Currently it is just SandDance.

Once SandDance is installed, when a user clicks on the visualizer icon, the SandDance extension will take over to load the query data in SandDance. This will be pushed in a separate pull request.

### CODE REVIEW

File: src/sql/workbench/parts/query/browser/queryResultsView.ts [hunk 1]
```diff
@@ -308,6 +307,8 @@ export class QueryResultsView extends Disposable {
 			dynamicTab.captureState(this.input.state.dynamicModelViewTabsState);
 		});
 
+		[this.resultsTab, this.messagesTab, this.qpTab, this.topOperationsTab, this.chartTab].forEach(t => t.clear());
```
Comment (kevcunnane) [inline src/sql/workbench/parts/query/browser/queryResultsView.ts]: What does this do / why was it added?

File: src/sql/workbench/parts/query/common/queryInput.ts [hunk 2]
```diff
@@ -253,6 +255,7 @@ export class QueryInput extends EditorInput implements IEncodingSupport, IConnec
 	public runQueryStatement(selection: ISelectionData): void {
 		this._queryModelService.runQueryStatement(this.uri, selection, this);
 		this.state.executing = true;
+
```
Comment (kevcunnane) [inline src/sql/workbench/parts/query/common/queryInput.ts]: nitpick: remove whitespace.

File: src/vs/platform/extensionManagement/common/extensionManagement.ts [hunk 3]
```diff
@@ -307,6 +307,8 @@ export interface IExtensionTipsService {
 
 	// {{SQL CARBON EDIT}}
 	getAppLaunchRecommendations(): Promise<IExtensionRecommendation[]>;
+	getVisualizerExtensions(): Promise<IExtensionRecommendation[]>;
+	promptVisualizerExtensions(): void;
```
Comment (kevcunnane) [inline src/vs/platform/extensionManagement/common/extensionManagement.ts]: Why does this need a prompt here, but no other case needs it?

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 4]
```diff
@@ -1132,5 +1132,87 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
...
+						localize('visualizer.VisualizerExtensions', "To utilize the Visualizer feature, the SandDance extension is required to be installed."),
+						[{
+							label: localize('visualizer.installAll', "Install Extension"),
+							run: () => {
+								/* __GDPR__
+								"extensionAppLaunchRecommendations:popup" : {
+									"userReaction" : { "classification": "SystemMetaData", "purpose": "FeatureInsight" }
+								}
+								*/
+								//this.telemetryService.publicLog('extensionAppLaunchRecommendations:popup', { userReaction: 'install' });
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: If this is commented, remove it and the GDPR tag, thanks!

File: src/vs/workbench/contrib/extensions/electron-browser/extensionsActions.ts [hunk 5]
```diff
@@ -1822,6 +1822,75 @@ export class InstallAppLaunchRecommendedExtensionsAction extends Action {
 	}
 }
 
+export class ShowVisualizerExtensionsAction extends Action {
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: Please move all of these to the src/sql tree.

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 6]
```diff
@@ -1132,5 +1132,87 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
 			.map(extensionId => (<IExtensionRecommendation>{ extensionId, sources: ['application'] })));
 	}
 
+	// called when Visualizer icon is clicked in the Query Results grid; "Never Show Again" is not an option
+	promptVisualizerExtensions(): void {
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: A few things:
- Any edit to the src/vs tree needs a SQL Carbon Edit comment block or it'll be deleted on next merge. Please verify any cases where you have this and add the comments.
- In this specific case, Let's not put this logic in the VS code area. There's no reason this couldn't just go directly in your action instead of in this service. I'd just move to the file you have the main action in for now.

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 7]
```diff
@@ -1132,5 +1132,87 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
...
+		let recommendations: IExtensionRecommendation[];
+		let localExtensions: ILocalExtension[];
+		const getRecommendationPromise = this.getVisualizerExtensions().then(recs => { recommendations = recs; });
+		const getLocalExtensionPromise = this.extensionsService.getInstalled(ExtensionType.User).then(local => { localExtensions = local; });
+		Promise.all([getRecommendationPromise, getLocalExtensionPromise]).then(() => {
+			if (!recommendations.every(rec => { return localExtensions.findIndex(local => local.identifier.id.toLocaleLowerCase() === rec.extensionId.toLocaleLowerCase()) !== -1; })) {
+				return new Promise<void>(c => {
+					this.notificationService.prompt(
+						Severity.Info,
+						localize('visualizer.VisualizerExtensions', "To utilize the Visualizer feature, the SandDance extension is required to be installed."),
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Given you have a visualizer category, why are you saying you need to install sandance here? Can you revise so it's accurate to the category?

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 8]
```diff
@@ -1132,5 +1132,87 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
...
+						[{
+							label: localize('visualizer.installAll', "Install Extension"),
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Would they really want to install all? I'm not sure we need this - let's discuss in the meeting.

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 9]
```diff
@@ -1132,5 +1132,87 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
...
+					);
+				});
+			} else {
+				return Promise.resolve();
+			}
+		});
+
+	}
+
+	getVisualizerExtensions(): Promise<IExtensionRecommendation[]> {
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: I would love to see if we can move all this code to SQL codebase. @alanrenmsft could you help advise on where to put this code?

File: src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+				viewlet.search('@visualizerExtensions');
+				viewlet.focus();
+			});
+	}
+}
+
+export class InstallVisualizerExtensionsAction extends Action {
+
+	static readonly ID = 'workbench.extensions.action.installVisualizerExtensions';
+	static LABEL = localize('installVisualizerExtensions', "Install Visualizer Extensions");
```
Comment (kevcunnane) [inline src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: @kburtram does this look OK to you? I'm happy pending some better suggestion from @alanrenmsft or @yualan and PM, but please confirm.

File: src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		this.recommendations = recommendations;
+	}
+
+	run(): Promise<any> {
+		return this.viewletService.openViewlet(VIEWLET_ID, true)
+			.then(viewlet => viewlet as IExtensionsViewlet)
+			.then(viewlet => {
+				viewlet.search('@visualizerExtensions');
+				viewlet.focus();
+				const names = this.recommendations.map(({ extensionId }) => extensionId);
```
Comment (kevcunnane) [inline src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: Let's cover case where this.recommendations is null since it's got to be set.

File: src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+				return this.extensionWorkbenchService.queryGallery({ names, source: 'install-visualizer-extensions' }, CancellationToken.None).then(pager => {
+					let installPromises: Promise<any>[] = [];
+					let model = new PagedModel(pager);
+					for (let i = 0; i < pager.total; i++) {
+						installPromises.push(model.resolve(i, CancellationToken.None).then(e =>
+							this.extensionWorkbenchService.install(e)
```
Comment (kevcunnane) [inline src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: You can put all this on the top line, or else should change to
```ts
e => {
    return this.extensionWorkbenchService.install(e);
}));
```
So that it's consistent with a multi-line function

File: src/sql/workbench/parts/query/browser/actions.ts [hunk 13]
```diff
@@ -199,13 +200,31 @@ export class ChartDataAction extends Action {
...
+	public static ID = 'grid.visualizer';
+	public static LABEL = localize('visualizer', 'Visualizer');
+	public static ICON = 'viewVisualizer';
+
+	constructor(@IEditorService private editorService: IEditorService, ) {
+		super(VisualizerDataAction.ID, VisualizerDataAction.LABEL, VisualizerDataAction.ICON);
+	}
+
+	public run(context: IGridActionContext): Promise<boolean> {
+		return Promise.resolve(true);
```
Comment (kevcunnane) [inline src/sql/workbench/parts/query/browser/actions.ts]: Is the logic for this in a separate PR?

File: src/sql/workbench/parts/query/browser/queryResultsView.ts (hunk 1 above)
Comment (kevcunnane) [reply src/sql/workbench/parts/query/browser/queryResultsView.ts]: Ping - why is this here?

File: src/sql/workbench/parts/query/common/queryInput.ts [hunk 14]
```diff
@@ -117,7 +118,8 @@ export class QueryInput extends EditorInput implements IEncodingSupport, IConnec
 		private _connectionProviderName: string,
 		@IConnectionManagementService private _connectionManagementService: IConnectionManagementService,
 		@IQueryModelService private _queryModelService: IQueryModelService,
-		@IConfigurationService private _configurationService: IConfigurationService
+		@IConfigurationService private _configurationService: IConfigurationService,
+		@IExtensionTipsService private _extensionTipsService: IExtensionTipsService
```
Comment (kevcunnane) [inline src/sql/workbench/parts/query/common/queryInput.ts]: Can you revert changes to this file? You're not using this constructor parameter anywhere

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 15]
```diff
@@ -1138,7 +1136,71 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
...
+		const getRecommendationPromise = this.getVisualizerExtensions().then(recs => { recommendations = recs; });
+		const getLocalExtensionPromise = this.extensionsService.getInstalled(ExtensionType.User).then(local => { localExtensions = local; });
+		Promise.all([getRecommendationPromise, getLocalExtensionPromise]).then(() => {
+			if (!recommendations.every(rec => { return localExtensions.findIndex(local => local.identifier.id.toLocaleLowerCase() === rec.extensionId.toLocaleLowerCase()) !== -1; })) {
+				return new Promise<void>(c => {
+					this.notificationService.prompt(
+						Severity.Info,
+						localize('visualizerRecommendations.VisualizerExtensions', "Azure Data Studio has data visualization extension recommendations."),
+						[{
+							label: localize('visualizerRecommendations.installAll', "Install Extension"),
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: This label is wrong. It should be "Install All Extensions" surely?

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 16]
```diff
@@ -1138,7 +1136,71 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
...
+							run: () => {
+								this.telemetryService.publicLog('visualizerRecommendations:popup', { userReaction: 'install' });
+								const installAllAction = this.instantiationService.createInstance(InstallVisualizerExtensionsAction, InstallVisualizerExtensionsAction.ID, localize('installAll', "Install All"), recommendations);
+								installAllAction.run();
+								installAllAction.dispose();
+								c(undefined);
+								const message = 'Your extension is ready. Select the Visualizer icon to visualize your data.';
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Please review this workflow with our UX / PM so it makes sense for multiple visualizers. the strings need review to support this.

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts [hunk 17]
```diff
@@ -1138,7 +1136,71 @@ export class ExtensionTipsService extends Disposable implements IExtensionTipsSe
...
+								this.notificationService.info(message);
+							}
+						}, {
+							label: localize('visualizerRecommendations.showMoreInfo', "More Info"),
```
Comment (kevcunnane) [inline src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Shouldn't this be "View Extensions" or similar? this all needs review with Alan Ren, UX & PM so it's consistent with other recommendations

File: src/sql/workbench/parts/query/browser/actions.ts (hunk 13 above)
Comment (RebeccaRunxinWang) [reply src/sql/workbench/parts/query/browser/actions.ts]: Yes it's in the PR I sent

File: src/vs/platform/extensionManagement/common/extensionManagement.ts (hunk 3 above)
Comment (kevcunnane) [reply src/vs/platform/extensionManagement/common/extensionManagement.ts]: OK, thanks!

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts (hunk 8 above)
Comment (kevcunnane) [reply src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Do the problem is you're not doing a specific extension anymore, it's a category. @angdesign and @alanrenmsft can you help here?

File: src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts (hunk 10 above)
Comment (ianyyu) [reply src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: @kevcunnane @yualan @kburtram 

Hi Kevin and all, the language we used here are following the MS writing style guide. According to the guide here are the principles we should follow: 
Our voice hinges on crisp simplicity. Bigger ideas and fewer words. Less head, more heart. 
**Our voice is:**
Warm and relaxed—We’re natural. Less formal, more grounded in real, everyday conversations. Occasionally, we’re fun. (We know when to celebrate.)
Crisp and clear—We’re to the point. We write for scanning first, reading the second. We make it simple above all.
Ready to lend a hand—We show customers we’re on their side. We anticipate their real needs and offer great information at just the right time.

**Example**
Replace this: Invalid ID
With this: You need an ID that looks like this: someone@example.com

Project friendliness
Use contractions: it’s, you’ll, you’re, we’re, let’s. To learn more, see Use contractions.

For all your reference, I have also attached the link to the guide. 
https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=28361
https://styleguides.azurewebsites.net/Styleguide/Read?id=2700
https://styleguides.azurewebsites.net/StyleGuide/Read?id=2696

Let me know if you have any other questions or concerns.

Comment (kevcunnane) [reply src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: Yep, the main thing here is being consistent with other extension recommendation language. That's been signed off & shipped in VSCode and Azure Data Studio. Karl had specifically asked to align with that in previous reviews & meetings, so I want this to be verified as being consistent.

File: src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts (hunk 11 above)
Comment (rchlkm) [reply src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: Thisclass, InstallVisualizerExtensionsAction, is only called when there are extensions to be recommended. This logic is handled in extensionTipsServices, and InstallVisualizerExtensionsAction is called in promptVisualizerExtensions(). 

Is the null case still necessary?

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts (hunk 6 above)
Comment (kevcunnane) [reply src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: No, it's all good.

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts (hunk 9 above)
Comment (kevcunnane) [reply src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: This is fine for now I think. Resolving.

File: src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts (hunk 11 above)
Comment (kevcunnane) [reply src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: General principle is to be self-contained. E.g. in this class you have a value that's not set in the constructor, but we "expected" to be there. Given this we would normally do null checking. I'm OK if you decide not to, it's more about defense in depth.

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts (hunk 8 above)
Comment (rchlkm) [reply src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Latest change: 
**Azure Data Studio has data visualization extension recommendations.**

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts (hunk 16 above)
Comment (kevcunnane) [reply src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Looks great. Resolving.

File: src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts (hunk 17 above)
Comment (kevcunnane) [reply src/vs/workbench/contrib/extensions/electron-browser/extensionTipsService.ts]: Awesome, thanks!

File: src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts (hunk 10 above)
Comment (rchlkm) [reply src/sql/workbench/contrib/extensions/electron-browser/extensionsActions.ts]: Updated "Install Visualizer Extensions" to "Install Extensions".

**Comment by rchlkm** [comment]
@kburtram @kevcunnane @alanrenmsft @angdesign @yualan @ianyyu  
Here are the most updated user facing strings that we believe is consistent with the current language used in ADS. Could we get a consensus? Thanks!

Visualization_popup

Visualization_popup2

File: src/sql/workbench/parts/query/browser/actions.ts [hunk 18]
```diff
@@ -199,13 +200,31 @@ export class ChartDataAction extends Action {
...
+	constructor(
+		@IEditorService private editorService: IEditorService,
+		@IExtensionTipsService private readonly extensionTipsService: IExtensionTipsService
+	) {
 		super(ChartDataAction.ID, ChartDataAction.LABEL, ChartDataAction.ICON);
 	}
 
 	public run(context: IGridActionContext): Promise<boolean> {
 		const activeEditor = this.editorService.activeControl as QueryEditor;
+		this.extensionTipsService.promptVisualizerExtensions();
```
Comment (kevcunnane) [inline src/sql/workbench/parts/query/browser/actions.ts]: One last thought: we should not start prompting on Stable (and, likely, even on Insiders) until an update to Sandance is officially published so users can actually benefit from data visualization in the query editor scenario. For now can we get this on product quality like we do for Jupyter Books? You can do in a separate PR if needed but this is a "master checkin" requirement.

```ts
import product from 'vs/platform/product/node/product';
// To filter out just stable
if (product.quality !== 'stable') {
    // prompt
}
// To filter out stable and insiders
if (product.quality !== 'stable' && product.quality !== 'insiders') {
    // prompt
}
```
